File: typescript/day24_part2_2021.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').split('\n');

let k: number[] = [];
let l: number[] = [];
let m: number[] = [];

input.forEach((line, i) => {
  let v: number;
  switch (i % 18) {
    case 4:
      v = Number(line.match(/div z (-?\d+)/)![1]);
      l.push(v);
      break;
    case 5:
      v = Number(line.match(/add x (-?\d+)/)![1]);
      k.push(v);
      break;
    case 15:
      v = Number(line.match(/add y (-?\d+)/)![1]);
      m.push(v);
      break;
  }
});

let constraints: { [key: number]: [number, number] } = {};
let stack: number[] = [];

l.forEach((v, i) => {
  switch (v) {
    case 1:
      stack.push(i);
      break;
    case 26:
      let pop = stack.pop()!;
      constraints[pop] = [i, m[pop] + k[i]];
      break;
  }
});

let min: number[] = new Array(14).fill(0);

for (let i = 0; i < 14; i++) {
  if (!(i in constraints)) continue;
  let vmin = 1;
  while (vmin + constraints[i][1] < 1) vmin++;
  min[i] = vmin;
  min[constraints[i][0]] = vmin + constraints[i][1];
}

console.log(num(min));

function num(w: number[]): number {
  let n = 0;
  for (let i of w) {
    n = n * 10 + i;
  }
  return n;
}

File: rust/day24_part2_2021.rs

use std::fs;
use std::str::FromStr;

fn main() {
    let content = fs::read_to_string("input.txt").expect("Unable to read file");
    let lines: Vec<&str> = content.trim().split('\n').collect();

    let mut k = Vec::new();
    let mut l = Vec::new();
    let mut m = Vec::new();

    for (i, line) in lines.iter().enumerate() {
        match i % 18 {
            4 => {
                let v = line.split_whitespace().nth(2).unwrap().parse::<i64>().unwrap();
                l.push(v);
            }
            5 => {
                let v = line.split_whitespace().nth(2).unwrap().parse::<i64>().unwrap();
                k.push(v);
            }
            15 => {
                let v = line.split_whitespace().nth(2).unwrap().parse::<i64>().unwrap();
                m.push(v);
            }
            _ => {}
        }
    }

    let mut constraints = std::collections::HashMap::new();
    let mut stack = Vec::new();

    for i in 0..l.len() {
        match l[i] {
            1 => {
                stack.push(i);
            }
            26 => {
                let pop = stack.pop().unwrap();
                constraints.insert(pop, (i, m[pop] + k[i]));
            }
            _ => {}
        }
    }

    let mut min = [0; 14];
    for i in 0..14 {
        if let Some(&(j, sum)) = constraints.get(&i) {
            let mut vmin = 1;
            while vmin + sum < 1 {
                vmin += 1;
            }
            min[i] = vmin;
            min[j] = vmin + sum;
        }
    }

    println!("{}", num(&min));
}

fn num(w: &[i64; 14]) -> i64 {
    let mut n = 0;
    for &digit in w {
        n = n * 10 + digit;
    }
    n
}
